### DISCUSSION TITLE
Draft functions

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#8528] [Ideas]
When you want to write a new function in Flix, you have to start by writing out a fully annotated signature. You could just give `Unit -> Unit` as the simplest annotation, but then you either have to update the signature each time it's required or accept that your WIP code will never compile. A solution is to support the concept of a draft function.

Annotated function definitions are good, but they limit the effectiveness of type inference when writing code.

# Ideal Solution
1 Write your function, will full inference of arguments
```scala
wip def add(x, y) = x + y
```
A `wip` function cannot be called anywhere, not even recursively (compiler should still track name to report duplicate names or better errors if you try to call `add`)
2 Annotate your function and remove `wip`
```scala
def add(x: a, y: a): a with Add[a] = x + y
```

## Design considerations
- Maybe all function definitions with some missing annotations are considered work in progress, thus removing the need for the `wip` modifier.
- Maybe all functions with `_` prefixed names are considered work in progress, thus removing the need for the `wip` modifier.
- Draft functions could also disable checks
  - about unused variables
  - about underspecified type classes (in a wip function `Add[alpha]` doesn't matter too much because alpha will perhaps be constrained later)

# Tasks to complete the dream
- Internal representation
  - [ ] add `wip` modifier
  - [ ] Track that `wip` variables are essentially unusable
- Parsing
  - [ ] Allow top level functions to omit annotations
  - [ ] Add parsing for `wip` modifier
- Wellformedness
  - [ ] Check that non-`wip` functions have annotations
- auto-complete / quick-fix
  - [ ] auto-add annotations of some subset of the function, the parameters individually, the return type, and the `wip` annotation

# Related things
- #6435
- comments of #3965

**Comment by mlutze** [comment]
Maybe we could also mark a module as `wip` to disable redundancy checks

**Comment by stetimi** [comment]
Would I be able to call a wip function from a test?

**Comment by JonathanStarup** [comment]
The infrastructure of the compiler is very based around functions being typable independently but calling a draft function would introduce a dependency which I think would require too much change
